src/core/handshaker/security/pipelined_secure_endpoint.cc

```diff
@@ -0,0 +1,938 @@
+        unprotecting_ = true;
+        return true;
+      }
+      return false;
+    }
+
+    static void FailReads(grpc_core::RefCountedPtr<Impl> impl,
+                          absl::Status status)
+        ABSL_LOCKS_EXCLUDED(frame_protector_.read_mu(), read_queue_mu_) {
+      impl->read_queue_mu_.Lock();
+      impl->unprotecting_ = status;
+      auto on_read = std::move(impl->on_read_);
+      impl->read_queue_mu_.Unlock();
+      impl.reset();
+      if (on_read != nullptr) on_read(status);
+    };
+
+    static void ContinueUnprotect(grpc_core::RefCountedPtr<Impl> impl) {
+      GRPC_LATENT_SEE_PARENT_SCOPE("secure endpoint continue unprotect");
+      ReadArgs args;
+      std::unique_ptr<SliceBuffer> source_buffer;
+      std::unique_ptr<SliceBuffer> read_buffer;
+      absl::Status unprotect_status;
+      absl::AnyInvocable<void(absl::Status)> on_read;
+
+      while (true) {
+        {
+          grpc_core::ReleasableMutexLock lock(&impl->read_queue_mu_);
+          if (!impl->unprotecting_.ok()) {
+            // Something failed or we're shutting down, so fail reads.
+            auto status = impl->unprotecting_.status();
+            lock.Release();
+            FailReads(std::move(impl), status);
+            return;
+          }
+
+          if (impl->protected_data_buffer_ == nullptr) {
+            // No pending data to unprotect - we're done.
+            impl->unprotecting_ = false;
+            lock.Release();
+            return;
+          }
+
+          // There's more data to unprotect - grab it under the queue lock.
+          source_buffer = std::move(impl->protected_data_buffer_);
+          impl->frame_protector_.TraceOp("data",
+                                         source_buffer->c_slice_buffer());
+          args = std::move(impl->last_read_args_);
+          args.set_read_hint_bytes(impl->frame_protector_.min_progress_size());
+        }
+
+        // Kick off the next read in another thread while we unprotect in this
+        // thread.
+        impl->event_engine_->Run(
+            [impl = impl->Ref(), args = std::move(args)]() mutable {
+              StartAsyncRead(std::move(impl), std::move(args));
+            });
+
+        {
+          grpc_core::ReleasableMutexLock lock(impl->frame_protector_.read_mu());
+          // Unprotect the bytes.
+          CHECK(read_buffer == nullptr);
+          impl->frame_protector_.SetSourceBuffer(std::move(source_buffer));
+          read_buffer = std::make_unique<SliceBuffer>();
+          impl->frame_protector_.BeginRead(read_buffer->c_slice_buffer());
+          unprotect_status = impl->frame_protector_.Unprotect(absl::OkStatus());
+          impl->frame_protector_.FinishRead(unprotect_status.ok());
+          if (!unprotect_status.ok()) {
+            lock.Release();
+            FailReads(std::move(impl), unprotect_status);
+            return;
+          }
+        }
+
+        {
+          grpc_core::ReleasableMutexLock lock(&impl->read_queue_mu_);
```

Comment:
Since we always call release manually maybe drop the addition scopes and just call read_queue_mu_.Lock()/Unlock() directly throughout?